lib/rspec/core/example_group.rb

```diff
@@ -769,6 +769,11 @@ def initialize(shared_group_name, inclusion_location)
         @inclusion_location = inclusion_location
       end
 
+      # @private
+      def ==(other)
+        @shared_group_name == other.shared_group_name
+      end
```

Comment:
> whilst we do allow groups to share the same name, they will never be in the same call stack.

I get that, but we don't know that only the recursion check will call `==` in the future.  If we're going to define a `==` method, I want it to have the correct equality semantics of this class.

Instead, WDYT about removing this method and doing this below?

``` ruby
def self.with_frame(name, location)
  current_stack = shared_example_group_inclusions
  if current_stack.any? { |shared_group| shared_group.name == name }
    raise ArgumentError, "can't include shared examples recursively"
  else
    current_stack << new(name, location)
    yield
  end
ensure
  current_stack.pop
end
```

Seems just as simple.  Does not rely on knowledge that `==` will be called (as @mrageh asked below).  And avoids defining a `==` method with semantics that don't actually match equality semantics.